src/Stage/InferTypes/AssignIds.elm

```diff
@@ -279,3 +279,137 @@ assignIdsWithHelp currentId located =
             in
             assignId newId <|
                 Typed.Record (Dict.fromList bindingBodiesList)
+
+        Canonical.Case e branches ->
+            let
+                ( e_, id1 ) =
+                    assignIdsWith currentId e
+
+                ( branches_, newId ) =
+                    List.foldr
+                        (\{ pattern, body } ( acc, runningId ) ->
+                            let
+                                ( typedPattern, bodyId ) =
+                                    assignPatternIdsWith runningId pattern
+
+                                ( typedBody, nextId ) =
+                                    assignIdsWith bodyId body
+                            in
+                            ( { pattern = typedPattern
+                              , body = typedBody
+                              }
+                                :: acc
+                            , nextId
+                            )
+                        )
+                        ( [], id1 )
+                        branches
+            in
+            assignId newId <|
+                Typed.Case e_ branches_
+
+
+assignPatternIdsWith : Id -> Canonical.LocatedPattern -> ( Typed.LocatedPattern, Id )
+assignPatternIdsWith currentId locatedCanonicalPattern =
+    let
+        ( typedPattern, newId ) =
+            assignPatternIdsWithHelp currentId (Located.unwrap locatedCanonicalPattern)
+    in
+    {- Keep location, for error context -}
+    ( Located.replaceWith typedPattern locatedCanonicalPattern
+    , newId
+    )
+
+
+assignPatternId : Id -> Typed.Pattern_ -> ( Typed.Pattern, Id )
+assignPatternId currentId located =
```

Comment:
Seems like this is the same as `assignId`, just with different type. Can we reuse it, and make it polymorphic instead? Or are the types making it impossible?